### ISSUE TITLE
tidy should not traverse untracked directories

### ISSUE THREAD

**Issue by pnkfelix** [rust-lang/rust#69291]
I ran (effectively):

```
% git pull origin master
...
% x.py test src/tools/tidy
...
<eventual complaint from tidy about files I have never touched>
```

There are often extraneous files and directories in my source tree. And often they are not even files/directories that I had created (see (*) below).

The problem is that tidy does a blanket traversal of the tree under `src/`, as far as I can tell, and thus will happily flag "problems" in files that are no longer effectively part of the source code. 

(My most recent instance of this was tidy flagging a problem in a `.rs` file under `src/stdsimd/`.)

Eventually, after several rounds of trying to understand why non-tidy code got checked into the github repo, I'll eventually remember to look at `git status` and take note of the directories listed under untracked files.  And then, usually, delete them.

But this should not be necessary. `tidy` should be smart enough to run `git status` itself and use that to drive its traversal.

----

(*) part of the above process is the build system will automatically update submodules. But old directories friom stick around,

**Comment by Mark-Simulacrum** [comment]
@pnkfelix Do you know which "part" of tidy is doing this? I suspect rustfmt myself, but am not quite sure :)

Ideally we'd do a global audit I guess for using `git ls-files` or so vs. direct traversal.

**Comment by pnkfelix** [comment]
No, I don't know which part is doing it. I was assuming it was `src/tools/tidy` that was gathering the set of paths to feed into `rustfmt`, but I have not confirmed that hypothesis.

**Comment by Mark-Simulacrum** [comment]
Well, I meant specifically what the errors were about -- I assume rustfmt based on your last message?

i.e., these are not (for example) line length or license check warnings, right?

**Comment by pnkfelix** [comment]
Oh I see.

I've only observed the problem with the `rustfmt` pass.

I just did a quick test; it seems like only the `rustfmt` checks are inspecting the untracked files. An untracked file with a long line does not seem to trigger the line length check.

I'm not clear on how the current license check works; I skimmed the `deps.rs` code in `src/tools/tidy` but couldn't work out where I should be finding a `vendor/` directory, which seems to be tied into the workings of the `tidy::deps::check` method.

**Comment by Mark-Simulacrum** [comment]
I spent a bit of time trying to look at this, but failed to quickly figure out how to get a list of files from git. AFAICT, `git ls-files` or `git ls-tree` don't quite do what we want. In particular, I was unable to coax git into listing checked in files but *not* to list src/test, for example.

It was my impression based on some searching online that this is perhaps even expected behavior. It's possible that we could do some sort of dual listing, where we both apply the current ignore rules, but also only visit files in that list.

I also checked, and it's my impression that all of our checks would hit this problem -- rustfmt is just significantly more likely to do so.

**Comment by pnkfelix** [comment]
> I also checked, and it's my impression that all of our checks would hit this problem -- rustfmt is just significantly more likely to do so.

This is true. (E.g. I just ran into it with untracked files that each had more than one trailing newine.)

**Comment by jyn514** [comment]
I suspect this was fixed by https://github.com/rust-lang/rust/pull/106440.

**Comment by jyn514** [comment]
Ok, the current behavior is that tidy doesn't look at *ignored* files (i.e. files in `.gitignore`), but it does look at *untracked* files. `x fmt --check` looks at neither.

The fix for this is probably to run `git ls-files . --exclude-standard --others` and skip files that it outputs.

**Comment by the8472** [comment]
Imo tidy _should_ check untracked files. If I run tidy before committing code and then commit+push I'd get CI errors because tidy skipped those.

**Comment by clubby789** [comment]
Removing assignment due to https://github.com/rust-lang/rust/pull/112921#issuecomment-1681164841

**Comment by heiseish** [comment]
I'm happy to open a PR in case no one is currently working on this, but I'm not quite sure what is the consensus here. 
I share the same opinion with @the8472, and it appears that the original problem can be avoided if tidy don't traverse submodules (Not sure if that's already the case so let me know if I'm missing something here)
Thanks!

**Comment by NoobProgrammer31** [comment]
assign it to me , i will look into it

**Comment by clubby789** [comment]
@NoobProgrammer31 Done 🙂 In future you can use rustbot to claim issues (see example [above](https://github.com/rust-lang/rust/issues/69291#issuecomment-1602404423) )
